base/set.jl

```diff
@@ -474,3 +474,249 @@ end
 
 convert(::Type{Set{T}}, s::Set{T}) where {T} = s
 convert(::Type{Set{T}}, x::Set) where {T} = Set{T}(x)
+
+
+
+## replace/replace! ##
+
+"""
+    replace!(A, old_new::Pair...; [count::Integer])
+
+For each pair `old=>new` in `old_new`, replace all occurrences
+of `old` in collection `A` by `new`.
+If `count` is specified, then replace at most `count` occurrences in total.
+See also [`replace`](@ref replace(A, old_new::Pair...)).
+
+# Examples
+```jldoctest
+julia> replace!([1, 2, 1, 3], 1=>0, 2=>4, count=2)
+4-element Array{Int64,1}:
+ 0
+ 4
+ 1
+ 3
+
+julia> replace!(Set([1, 2, 3]), 1=>0)
+Set([0, 2, 3])
+```
+"""
+replace!(A, old_new::Pair...; count::Integer=typemax(Int)) = _replace!(A, eltype(A), count, old_new)
+
+# we use this wrapper because using directly eltype(A) as the type
+# parameter below for Some degrades performance
+function _replace!(A, ::Type{K}, count::Integer, old_new::Tuple{Vararg{Pair}}) where K
+    @inline function prednew(x)
+        for o_n in old_new
+            first(o_n) == x && return Some{K}(last(o_n))
+        end
+    end
+    replace!(prednew, A, count=count)
+end
+
+"""
+    replace!(pred::Function, A, new; [count::Integer])
+
+Replace all occurrences `x` in collection `A` for which `pred(x)` is true
+by `new`.
+
+# Examples
+```jldoctest
+julia> A = [1, 2, 3, 1];
+
+julia> replace!(isodd, A, 0, count=2)
+4-element Array{Int64,1}:
+ 0
+ 2
+ 0
+ 1
+```
+"""
+replace!(pred::Callable, A, new; count::Integer=typemax(Int)) =
+    replace!(x -> if pred(x) Some(new) end, A, count=count)
```

Comment:
Won't this be `nothing` when `pred(x)` is `false`? That seems oddly implicit. IMO it would be clearer to write it as `x->pred(x) ? Some(new) : nothing` if that is indeed the intent.

Reply:
I find lovely this implicit `nothing`, but I appreciate many will disagree. I was wondering if this is a pattern which will emerge, to pune on the `nothing` return value from `if` without `else` and `for`. I remember that you precisely would have prefered something else than `Void` to replace `Nullable`... I'm absolutely fine to change this here, but will do next time there is a more serious reason to push (CI resources are scarces, though not as much as on deadline day ;-) )

Reply:
I don't think it matters much and it's kind of cute that this works, but I have to say I do prefer the explicitness of `pred(x) ? Some(new) : nothing`. The `if` version requires the reader to know that `if` without and `else` evaluates to `nothing` if the condition is false.